### ISSUE TITLE
Missing sessions and passwords after upgrade with Chromium changeover

### ISSUE THREAD

**Issue by ylluminate** [dmarmor/epichrome#203]
As others have noted, I'm now also having the Chromium missing sessions / no saved passwords issues... I'm going to hold off on upgrading any more Epichrome apps until we know we've got this sorted out.

Please let me know where you'd like for me to go with this. Do you think that perhaps the app/engine consolidation will help and do you want to try that first or do we need to work on isolating this separately?

**Comment by dmarmor** [comment]
Thanks for starting this threat. @henderea first experienced this, which is an issue where for some users and some apps, the Chromium engine for currently unknown reasons refuses to request permission to use the Chromium Safe Storage item in the user's system keychain. It may have started with Chromium 79.0.3945.130, as @henderea never had the problem with 2.3.0b1, which used 79.0.3945.117.

Currently I can't reproduce it on my system, so have no idea how to fix it. I'm kind of hoping it'll fix itself with the next Chromium build. In the meantime, I'm going to proceed with the app/engine consolidation (which will address issue #201 as well as another problem @henderea noticed where 1Password wasn't working even with the Chrome engine--there's currently no Github issue for that). Once that's done (and maybe there's a new Chromium build) I think we should see if it helps. If not, we can work more on diagnosing.

**Comment by ylluminate** [comment]
Sounds good. I am anxiously awaiting the next build that integrates the consolidation to see if it helps. I'm in a holding pattern with updating Epichrome any more until this is released. Do you think we will see this consolidation build drop soon?

**Comment by dmarmor** [comment]
Just posted the new beta. I'm not super otpimistic moving the engines is going to help with this issue, but at least it's a starting point. Once you have Chromium apps running with beta 4, if you experience the problem, let me know and maybe we can start to diagnose it...

**Comment by ylluminate** [comment]
So just to find my feet here and make sure things are clear: The new installer creates `/Applications/Epichrome` that contains these, right?:
• `Epichrome.app`
• `EpichromeEngines.noindex`

Currently all 3 of these folders are owned by root:wheel. Can these be chowned to $USER:staff or $USER:wheel without negative repercussions? My thought is to at least have `/Applications/Epichrome` owned by $USER and then I'll just push ALL of my Epichrome apps from `~/Applications` and `/Applications` into it and then work on upgrading systematically...

Or is there a better way to get started down this road?

As I do this, of course, I'm going to watch for this problem from this issue and we'll systematically (hopefully) figure out what's going on...

**Comment by dmarmor** [comment]
Yeah, you can chown the outer folder no problem. Really, if you're the only user on the system, you can chown it all if you want. It's just the way the package installer works that it makes it all root-owned anyway. The one good feature of that is that the EpichromeEngines.noindex folder is owned by root and set sticky, so that users can't accidentally (or maliciously) remove other users' engines. But if you, like me, are on a single-user system, then chown away. It shouldn't affect anything.

**Comment by ylluminate** [comment]
So some notes on / with upgrade:

- During upgrade I got a dialogue that said: `Warning: Unable to check for new version of Epichrome on GitHub: `
- I tried saving a password and it is still not in settings. It neither appears in `settings > passwords` when I first create it nor when I restart the Epichrome app.

**Comment by dmarmor** [comment]
Hmm, that's odd about checking Github. You didn't by chance save the log file from that run, did you? It'd be useful to me to see if I can see what error caused that warning. If not, I can talk you through an edit to config.sh that will cause it to try to check again on the next run, and then would love to have the log file from that run.

In terms of saving a password, this is a Chromium-engine app, I assume? If so, would you mind doing some tests to help diagnose? The downside of these tests is they will kill any saved sessions you have in any of your Chromium apps. If that's problematic for you, another way to do it would be to create a new system user and log into that user to do the tests. A pain, of course, but it would help me figure out what's going on. Anyway, let me know if you'd be able to do any of that...

**Comment by dmarmor** [comment]
Unfortunately, with the current architecture there's no good way to stay connected to the Chromium terminal output. Running the executable directly was how I used to launch in the old architecture, but with Catalina it caused all manner of problems, so I switched to launching using /usr/bin/open. It's pretty rock-solid, but unfortunately doesn't give us terminal output.

**Comment by ylluminate** [comment]
Can we use direct execution temporarily to isolate output? I assume that I could modify the Epichrome script at some point to trigger this old behavior to just see it for myself?

**Comment by dmarmor** [comment]
Any direct execution of the engine executable causes it not to launch properly in Catalina, so I don't think that will work. If you're really determined, this method might work, but it's pretty ugly (involving using gdb):

https://stackoverflow.com/questions/593724/redirect-stderr-stdout-of-a-process-after-its-been-started-using-command-lin

related:
https://etbe.coker.com.au/2008/02/27/redirecting-output-from-a-running-process/

**Comment by ylluminate** [comment]
Curious too, have you played with `dtruss` attachment?

**Comment by dmarmor** [comment]
Nope! I don't even know what dtruss is. A quick glance at the man page was a bit confusing. What does it do?

**Comment by ylluminate** [comment]
I used to use `strace` in Linux and `ktrace` before Apple yanked it out in favor of `dtrace`. These were useful before to trace out system calls to figure out what was going on in an application or process as far as seeing all system calls that were flying around. Apple began to obfuscate this process (I assume at the behest of commercial vendors) and so it has become a bit more difficult to sort through this stuff unless you use some obfuscation defeating tools (contact me off issue to discuss details). 

At any rate, it should be possible to par down the verbosity to the point where you just see things that are relevant to certain filesystem paths that might have to do with password saving. The trick is what items to eliminate and thus this is where the complexity increases here. 

I suspect the gdb method might be simpler actually so I'll look at it first. At the end of the day I actually wouldn't mind getting my hands around this entirely since I have a feeling it will also reveal other problems I've been facing that I've mentioned before.

**Comment by ylluminate** [comment]
Just to document what I'm up to: So I thought I'd poke about in `gdb` to see if it was straight forward. I guess I'm going to have to dig into this a bit deeper since the current method outlined in the links above is not working:
```
sudo gdb -p <PID>
# blah blah
Attaching to process <PID>
[New Thread 0xa03 of process <PID>]
[New Thread 0xb03 of process <PID>]
[New Thread 0xc03 of process <PID>]
[New Thread 0x1503 of process <PID>]
[New Thread 0x1603 of process <PID>]
[New Thread 0x1703 of process <PID>]
[New Thread 0x1803 of process <PID>]
[New Thread 0x1903 of process <PID>]
[New Thread 0x1a03 of process <PID>]
[New Thread 0x1b03 of process <PID>]
[New Thread 0x1c03 of process <PID>]
[New Thread 0x1d03 of process <PID>]
[New Thread 0x1e03 of process <PID>]
[New Thread 0x1f03 of process <PID>]
[New Thread 0x2003 of process <PID>]
[New Thread 0x2103 of process <PID>]
[New Thread 0x2203 of process <PID>]
[New Thread 0x2303 of process <PID>]
[New Thread 0x2403 of process <PID>]
[New Thread 0x2503 of process <PID>]
[New Thread 0x2603 of process <PID>]
[New Thread 0x2703 of process <PID>]
[New Thread 0x2803 of process <PID>]
[New Thread 0x2903 of process <PID>]
[New Thread 0x2a03 of process <PID>]
[New Thread 0x2b03 of process <PID>]
[New Thread 0x5403 of process <PID>]
Error calling thread_get_state for GP registers for thread 0xa03

warning: Mach error at "i386-darwin-nat.c:132" in function "virtual void i386_darwin_nat_target::fetch_registers(struct regcache *, int)": (os/kern) invalid argument (0x4)
Reading symbols from /Applications/Epichrome/EpichromeEngines.noindex/USERNAME/APP_NAME/APP_NAME.app/Contents/MacOS/Chromium...
(No debugging symbols found in /Applications/Epichrome/EpichromeEngines.noindex/USERNAME/APP_NAME/APP_NAME.app/Contents/MacOS/Chromium)

warning: unhandled dyld version (16)
0x00007fff6ef3c25a in ?? ()
(gdb) p dup2(open("/dev/null",0),1)
No symbol table is loaded.  Use the "file" command.
(gdb) p dup2(open("/dev/null",0),2)
No symbol table is loaded.  Use the "file" command.
```

**Comment by ylluminate** [comment]
So just a thought here @dmarmor, would you like to see a packed up app that is exhibiting this no-password-save behavior to see if you see anything yourself?

**Comment by dmarmor** [comment]
Yes, please! That could be helpful.

**Comment by ylluminate** [comment]
Done.

**Comment by ylluminate** [comment]
Sorry, beta 5 did not help the password saving situation – however I did notice that my session was saved this time when I opened it up... So perhaps this is now limited to passwords only for me being lost / not saving...

**Comment by dmarmor** [comment]
Can you clarify? If you run an app that's updated to beta 5 and log into a website, then quit and run again, does it return to the session logged in? If that's the case, then the problem is indeed fixed, even if it doesn't have your old session data. (I've been using github.com for this testing actually, as I know it should save my login across a quit.)

**Comment by ylluminate** [comment]
Well, I can't say I was careful in testing the session existence prior. I thought it was not working, BUT I frankly cannot recall and I did not document it well enough due to my hectic schedule over the last week.  Yes, the session aspect may well indeed be solved.

I did test the password saving and it did NOT work.  The password issue has been more important to me and thus I was more focussed on that...

I think we're going to need @henderea to chime in on a session saving differential / resolution confirmation...

**Comment by ylluminate** [comment]
Beta 6 did bring back passwords. I loaded up the app, logged in and then saved the password. Before saving the password old passwords did not appear, but after saving the password this time all of the old saved passwords appeared this time as well that had been saved in the previous Chrome version of Epichrome and of course quitting and restarting the app ALSO revealed all apps yet again.

One differentiating thing that did happen this time (because it was Chrome I assume) was that I was prompted for login: 

![example](https://p153.p0.n0.cdn.getcloudapp.com/items/L1ugGlgA/keychain.jpg)

**Comment by dmarmor** [comment]
OK, that is very promising! That prompt is good news, as that's what I've gotten on my system both with the old Chromium betas & this experimental version, where I've been able to save persistent data across runs. So it seems like this version might well be solving at least that problem...

**Comment by dmarmor** [comment]
Hey guys, I believe this has been resolved as well as possible in 2.3.0. There are lots of warnings now when switching engines that passwords and sessions will be lost, and password importing is working. Closing this issue, but as usual, we can reopen if and when we need to.
